drift-transport-netty/src/test/java/com/facebook/drift/transport/netty/server/TestDriftNettyServerTransport.java

```diff
@@ -190,14 +190,15 @@ private static int testOutOfOrderNotSupported(
                 // send second request, which will be blocked in the server because this client does not support out of order responses
                 // the only way to test this is with a sleep because the request is blocked inside of the server IO stack
                 sendLogRequest(22, messages, protocol);
-                assertNull(results.poll(1, SECONDS), "Second request future");
+                SettableFuture<Object> secondResult = results.poll(5, SECONDS);
+                assertNotNull(secondResult, "expected a future to be enqueued for second request");
+                assertFalse(secondResult.isDone());
```

Comment:
I took a deeper look at this change.

Imagine we have two requests.
The first one is a db read (some network call or a heavy weight op) and second is just an echo.

The fact that the second request future is not null as in the previous code is problematic.
In this test case the results are set from the client side and not from the server side and hence `secondResult` is never done until a value is set later.
Note that the `results` queue is populated from `com.facebook.drift.transport.netty.server.TestDriftNettyServerTransport.TestingServerMethodInvoker#invoke`

This means that server already started processing the second request with this new change.

From a small test I coded up to illustrate what is happening.

Sequence of events with old code.
```
Sent DB request
Fire Channel Read  <---- from ResponseOrderingHandler
Added SettableFuture for DB in result queue
Sent Echo request
/-/-/-/-/-/-/-/-/-/-/-/-/-/-
Async : Waiting for DB to finish <------ Waits for some time
/-/-/-/-/-/-/-/-/-/-/-/-/-/-
DB done
Fire Channel Read <----  Second req. This is only fired after the DB read is done.
Added SettableFuture for Echo in result queue
Echo done
Sending Echo response back
```

Sequence of events with new code.
```
Sent DB request
Fire Channel Read
Added SettableFuture for DB in result queue
Sent Echo request
/-/-/-/-/-/-/-/-/-/-/-/-/-/-
Async : Waiting for DB to finish
/-/-/-/-/-/-/-/-/-/-/-/-/-/-
Fire Channel Read <---- This is fired before the DB read is done.
Added SettableFuture for Echo in result queue
Echo done  <----  Echo is done before the DB read is done.
```

Reply:
Good catch, thanks for the test case! 

I don't think `ByteToMessageDecoder`, superclass of `ThriftProtocolDetection`, respects the auto read config the way we want it to. I'm also unsure if it's a fluke that this worked before that ["blame" change](https://github.com/netty/netty/commit/6b6475fb565551c36f372fe37683d98df6674161 ) in [4.1.38.Final`](https://netty.io/news/2019/07/24/4-1-38-Final.html).

I think `FlowControlHandler` is an appropriate addition here. Its [javadoc](https://netty.io/4.1/api/io/netty/handler/flow/FlowControlHandler.html) states

> [plexHandler](https://netty.io/4.1/api/io/netty/channel/ChannelDuplexHandler.html)
The [FlowControlHandler](https://netty.io/4.1/api/io/netty/handler/flow/FlowControlHandler.html) ensures that only one message per read() is sent downstream. Classes such as [ByteToMessageDecoder](https://netty.io/4.1/api/io/netty/handler/codec/ByteToMessageDecoder.html) or [MessageToByteEncoder](https://netty.io/4.1/api/io/netty/handler/codec/MessageToByteEncoder.html) are free to emit as many events as they like for any given input. A channel's auto reading configuration doesn't usually apply in these scenarios. This is causing problems in downstream [ChannelHandler](https://netty.io/4.1/api/io/netty/channel/ChannelHandler.html)s that would like to hold subsequent events while they're processing one event.

This describes our use case pretty well. In fact, the source code also describes the typical use case as

```
 * class MyExampleHandler extends ChannelInboundHandlerAdapter {
 *   @Override
 *   public void channelRead(ChannelHandlerContext ctx, Object msg) {
 *     if (msg instanceof HttpRequest) {
 *       ctx.channel().config().setAutoRead(false);
 *
 *       // The FlowControlHandler will hold any subsequent events that
 *       // were emitted by HttpObjectDecoder until auto reading is turned
 *       // back on or Channel#read() is being called.
 *     }
 *   }
 * }
 * }</pre>
```
which is exactly what `ResponseOrderingHandler` already does.

So I'm adding this handler to the Thrift pipeline before `ResponseOrderingHandler` I'm also updating the version of Netty a bit further.

Reply:
I couldn't really repro this call order, but I was seeing the 
```
expected sequenceId 11, but received 22
```
assertion failures.

Turns out the call to 
```
context.channel().config().setAutoRead(true);
```
actually triggers the read inline, which might lead to a full request-response cycle happening before the call to 
```
context.write(message, promise);
```
that would follow. So you'd get the response for the read that happened after reads were re-enabled getting written before the response from before.

Changing the order so that the `write` happens before `setAutoRead(true)` should fix things. 

Updated PR.